keps/sig-instrumentation/4828-component-flagz/README.md

```diff
@@ -0,0 +1,388 @@
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+We propose to enhance the observability, troubleshooting, and debugging capabilities of core Kubernetes components by implementing a flagz endpoint for all components. This will provide users with real-time visibility into the command-line flags each component was started with, aiding in the diagnosis of configuration issues and unexpected behavior
+
+## Motivation
+
+This proposal extends the ideas presented in the [Component Statusz KEP](https://github.com/kubernetes/enhancements/pull/4830), however, it proposes a distinct flagz endpoint to enhance visibility into component configurations. By providing real-time visibility into the active configuration of Kubernetes components, flagz enables users to quickly spot misconfigurations that could lead to system instability or outages
+
+### Goals
+
+* Provide users with the ability to dynamically inspect and understand the active flags of running Kubernetes components
+
+### Non-Goals
+
+* Replace existing monitoring solutions: metrics, logs, traces
+* Provide information about components that are not directly accessible due to network restrictions
+
+## Proposal
+
+We are proposing to add a new endpoint, /flagz on all core Kubernetes components which returns the command-line flags each component was started with.
+
+### Risks and Mitigations
+
+1. **No sensitive data exposed**
+    
+    We will ensure that no sensitive data is exposed through flagz and that access to this endpoint is gated by using system-monitoring group.
+
+2. **Performance impact**
+    
+    We will take care to carefully consider the computational overhead associated with the flagz endpoint to avoid negatively impacting the server's overall performance.
+
+3. **Essential data only**
+    
+    We will only display the command-line flag values for a component in this page and nothing more.
+
+4. **Premature Dependency on Unstable Format**
+
+    The alpha release will explicitly support plain text format only, making it clear that the output is not intended for parsing or automated monitoring. The feature will be secured behind a feature gate that is disabled by default, ensuring users opt-in consciously. Also, the endpoint will be implemented with versioned URLs to facilitate future transitions to structured formats without breaking existing integrations, once the usage patterns and requirements are better understood.
+
+## Design Details
+
+### Data Format and versioning
+
+Initially, the flagz page will exclusively support a plain text format for responses. We will implement these endpoints using versioned URLs, like /v1/flagz, to ensure future compatibility. This versioning strategy allows us to seamlessly introduce structured data formats (e.g., JSON) in the future, through a distinct endpoint such as /v2/flagz, without disrupting existing implementations.
```

Comment:
Wouldn't it be better to have the endpoints in the form of `/flagz/v1`? Flagz is the subdomain that we want to version, if we were to put the version first, that would be confusing with the component version as it would then be at the top-level.

Reply:
Yes you're right, I got that the other way round. Fixed it now, thanks!

Reply:
One thing I was wondering since then is whether going with a query parameter might not be better. Something like:

```
/flagz?version=1
/flags?version=2
```

wdyt?

Reply:
This was also raised in the [statusz KEP PR](https://github.com/kubernetes/enhancements/pull/4830#discussion_r1783541095). I am ok with either, whichever seems the best to indicate the version for the endpoint.

Reply:
Updated the KEP to use query params for specifying the version.